core/coretypes/src/struct_builder_impl.cpp

```diff
@@ -0,0 +1,179 @@
+#include <coretypes/struct_builder_impl.h>
+#include <coretypes/validation.h>
+#include <coretypes/struct_factory.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+    StructBuilderImpl::StructBuilderImpl(const StringPtr& name, const TypeManagerPtr& typeManager)
+    : fields(Dict<IString, IBaseObject>())
+{
+    structType = typeManager.getType(name);
+
+    const auto defaultValues = structType.getFieldDefaultValues();
+    const auto names = structType.getFieldNames();
+    
+    for (size_t i = 0 ; i < names.getCount(); ++i)
+    {
+        if (defaultValues.assigned())
+            fields.set(names[i], defaultValues[i]);
+        else
+            fields.set(names[i], nullptr);
+    }
+}
+
+StructBuilderImpl::StructBuilderImpl(const StructPtr& struct_)
+    : fields(Dict<IString, IBaseObject>())
+{
+    structType = struct_.getStructType();
+    for(const auto& field : struct_.getAsDictionary())
+        fields.set(field.first, field.second);
+}
+
+ErrCode StructBuilderImpl::build(IStruct** struct_)
+{
+    OPENDAQ_PARAM_NOT_NULL(struct_);
+
+    const auto builderPtr = this->borrowPtr<StructBuilderPtr>();
+
+    return daqTry([&]()
+    {
+        *struct_ = StructFromBuilder(builderPtr).detach();
+        return OPENDAQ_SUCCESS;
+    });
+}
+
+ErrCode StructBuilderImpl::setFieldValues(IList* values)
+{
+    OPENDAQ_PARAM_NOT_NULL(values);
+
+    const auto& valuesPtr = ListPtr<IString>::Borrow(values);
+    if (valuesPtr.getCount() != fields.getCount())
+        return OPENDAQ_ERR_INVALIDPARAMETER;
+
+    const auto types = structType.getFieldTypes();
+    for (size_t i = 0; i < valuesPtr.getCount(); ++i)
+        if (valuesPtr.assigned() && SimpleType(valuesPtr[i].getCoreType()) != types[i])
+            return OPENDAQ_ERR_INVALIDPARAMETER;
+
+    DictPtr<IString, IBaseObject> newDict = Dict<IString, IBaseObject>();
+    const auto keyList = fields.getKeyList();
+    for (size_t i = 0; i < valuesPtr.getCount(); ++i)
+        newDict.set(keyList[i], valuesPtr[i]);
+
+    fields = newDict.detach();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode StructBuilderImpl::set(IString* name, IBaseObject* field)
+{
+    OPENDAQ_PARAM_NOT_NULL(name);
```

Comment:
not consistent in this file, somewhere OPENDAQ_PARAM_NOT_NULL and somewhere "if" statement